File: colcon_ws/src/my_cpp_pkg/src/my_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

#include "geometry_msgs/msg/twist.hpp"

#include "turtlesim/msg/pose.hpp"
#include <cmath>
#include <optional>

//---------------------------------------------------------------

class MinimalPublisher: public rclcpp::Node
{
private:
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    int count_;

    void timer_callback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello " + std::to_string(count_++);
        publisher_->publish(message);
        RCLCPP_INFO(this->get_logger(), "Pusblishing:'%s'", message.data.c_str());
    }

public:
    MinimalPublisher():Node("minimal_publisher"), count_(0)
    {
        publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&MinimalPublisher::timer_callback, this)
        );
    }
};

//---------------------------------------------------------------

class TurtleController:public rclcpp::Node
{
private:
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;

    void timer_callback()
    {
        auto msg = geometry_msgs::msg::Twist();
        msg.linear.x = 2.0;
        msg.angular.z = 1.0;
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Pusblished: linear=%.1f, angular=%.1f", 
                    msg.linear.x, msg.angular.z);
    }

public:
    TurtleController():Node("tutle_controller")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&TurtleController::timer_callback, this)
        );
    }
};

//---------------------------------------------------------------

class SpiralPublisher: public rclcpp::Node
{
private:
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    int count = 0;

    void timer_callback()
    {
        auto msg = geometry_msgs::msg::Twist();
        msg.linear.x = 2.0 + count * 0.05;
        msg.angular.z = 1.0;
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Pusblished: linear=%.1f, angular=%.1f", 
                    msg.linear.x, msg.angular.z);

        count += 1;
    }

public:
    SpiralPublisher():Node("spiral_controller")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&SpiralPublisher::timer_callback, this)
        );
    }
};

//---------------------------------------------------------------

class EightFigureController: public rclcpp::Node
{
private:
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscriber_;
    rclcpp::TimerBase::SharedPtr timer_;
    int count = 0;
    float linear_speed = 1.0;
    float angular_speed = 2.0;
    float current_angular_z = angular_speed;
    std::optional<float> last_theta;
    float cumulative_angle = 0.0;

    void timer_callback()
    {
        auto msg = geometry_msgs::msg::Twist();
        msg.linear.x = linear_speed;
        msg.angular.z = current_angular_z;
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Pusblished: linear=%.1f, angular=%.1f", 
                    msg.linear.x, msg.angular.z);

        count += 1;
    }

    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        float theta = msg->theta;

        if (last_theta)
        {
            float delta = theta - *last_theta;
            delta = std::fmod(delta + M_PI, 2 * M_PI);
            if (delta < 0.0)
            {
                delta += 2 * M_PI;
            }
            delta -= M_PI;

            cumulative_angle += delta;

            if (std::fabs(cumulative_angle) > 2 * M_PI)
            {
                current_angular_z *= -1.0;
                cumulative_angle = 0.0;
                RCLCPP_INFO(this->get_logger(), "--> Change Direction !!");
            }
        }

        last_theta = theta;
    }

public:
    EightFigureController():Node("eight_figure_controller")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        subscriber_ = this->create_subscription<turtlesim::msg::Pose>("turtle1/pose", 10,
            std::bind(&EightFigureController::pose_callback, this, std::placeholders::_1));
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&EightFigureController::timer_callback, this)
        );
    }
};



int main(int argc, char* argv[])
{
    int test_type = 4;
    rclcpp::init(argc, argv);

    if (test_type == 1)
    {
        rclcpp::spin(std::make_shared<MinimalPublisher>());
    }
    else if (test_type == 2)
    {
        rclcpp::spin(std::make_shared<TurtleController>());
    }
    else if (test_type == 3)
    {
        rclcpp::spin(std::make_shared<SpiralPublisher>());
    }    
    else if (test_type == 4)
    {
        rclcpp::spin(std::make_shared<EightFigureController>());
    }       
    else
    {
        rclcpp::spin(std::make_shared<MinimalPublisher>());
    }
        
    rclcpp::shutdown();
    return 0;
}

File: colcon_ws/src/my_py_pkg/my_py_pkg/my_node.py
"""
1秒ごとに、 Hello N と chatter という topic に送信し続ける publisher node 
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
import math

class MinimalPublisher(Node):
    """
    定常円運動
    """
    def __init__(self):
        # Nodeクラスの初期化
        super().__init__('minimal_publisher')
        
        # String を chatter というトピックに送る publisherを作成. buffer size = 10
        self.publisher_ = self.create_publisher(String, 'chatter', 10)
        
        # 1秒間隔で、 自作関数 self.timer_callback を呼び出す timer を作成
        timer_period = 1.0
        self.timer = self.create_timer(timer_period, self.timer_callback)
        
        # カウンターの初期化
        self.count = 0
    
    # 1秒ごとに呼ばれる自作関数
    def timer_callback(self):
        msg = String()
        msg.data = f'Hello {self.count}'
        self.publisher_.publish(msg)
        self.get_logger().info(f'Pusblishing: "{msg.data}"')
        self.count += 1


class TurtleController(Node):
    """
    螺旋を描く
    """
    def __init__(self):
        super().__init__('turtle_controller')
        self.publisher_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.count = 0
    
    def timer_callback(self):
        msg = Twist()
        msg.linear.x = 2.0 + self.count*0.05 # 少しずつ速度を上げることで半径を大きくする v=r*w
        msg.angular.z = 1.0
        self.publisher_.publish(msg)
        self.get_logger().info(f'Pusblished: linear={msg.linear.x}, angular={msg.angular.z}')
        
        self.count += 1


class EightFigureController(Node):
    """
    8の字を描く
    """
    def __init__(self):
        super().__init__('eight_figure_controller')
        
        self.linear_speed = 1.0
        self.angular_speed = 2.0
        self.current_angular_z = self.angular_speed
        
        self.last_theta = None
        self.cumulative_angle = 0.0
        
        self.publisher_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.subscriber_ = self.create_subscription(Pose, '/turtle1/pose', self.pose_callback, 10)
        
        self.timer = self.create_timer(0.1, self.timer_callback)
    
    def pose_callback(self, msg):
        theta = msg.theta
        
        # 差分を計算し、-π〜+πにラップ
        if self.last_theta is not None:
            delta = theta - self.last_theta
            delta = (delta + math.pi) % (2 * math.pi) - math.pi # -pi to +pi
            self.cumulative_angle += delta
            
            # 1回転（約2π）以上で方向反転
            if abs(self.cumulative_angle) > 2 * math.pi:
                self.current_angular_z *= -1
                self.cumulative_angle = 0.0
                self.get_logger().info('--> Inverse Direction')

        self.last_theta = theta
        
    def timer_callback(self):
        msg = Twist()
        msg.linear.x = self.linear_speed
        msg.angular.z = self.current_angular_z
        self.publisher_.publish(msg)
        
def main(args=None):
    test_type = 3
    rclpy.init(args=args)
    #
    if test_type == 1:
        node = MinimalPublisher()
    elif test_type == 2:
        node = TurtleController()
    elif test_type == 3:
        node = EightFigureController()
    else:
        node = MinimalPublisher()
    #
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()